### PR TITLE
ARROW-7819: [C++][Gandiva] Add DumpIR to Filter/Projector object

### PR DESCRIPTION
**Pull request by fsaintjacques**
The following patch exposes the generated IR as a method of the objects
for further inspection. This is a breaking change for the internal
method `FinalizeModule` which doesn't take the dump_ir and optimize
flags, it receives `optimize` from Configuration now.

- Refactored Engine, notably removed dead code, organized init in a single
  function and simplified LLVMGenerator.
- Dumping IR should not write to stdout, but instead return it as a
  string in the `DumpIR` method.
- Refactored Types, fixing some bad methods type.
- Added the optimize field to `Configuration` class.
- Simplified some unit tests.

But more importantly, we can now inspect dynamically:

```python
>>> filter = gandiva.make_filter(table.schema, condition)                                                                                    
>>> print(filter.ir)
; ModuleID = 'codegen'                                                
source_filename = "codegen"                    
target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"           
target triple = "x86_64-unknown-linux-gnu"                 
                                                                      
@llvm.global_ctors = appending global [0 x { i32, void ()*, i8* }] zeroinitializer
@_ZN5arrow7BitUtilL8kBitmaskE = internal unnamed_addr constant [8 x i8] c"\01\02\04\08\10 @\80", align 1
                                                                                                                                             
; Function Attrs: norecurse nounwind                      
define i32 @expr_0_0(i64* nocapture readonly %args, i64* nocapture readonly %arg_addr_offsets, i64* nocapture readnone %local_bitmaps, i16* nocapture readnone %selection_vector, i64 %context_ptr, i64 %nrecords) local_unnamed_addr #0 {
entry:                                                                
  %0 = bitcast i64* %args to i8**                                                                                                            
  %cond_mem56 = load i8*, i8** %0, align 8                            
  %1 = getelementptr i64, i64* %arg_addr_offsets, i64 3            
  %2 = load i64, i64* %1, align 8                                                                                                            
  %a_mem_addr = getelementptr i64, i64* %args, i64 3                  
  %3 = bitcast i64* %a_mem_addr to double**            
  %a_mem7 = load double*, double** %3, align 8                        
  %scevgep = getelementptr double, double* %a_mem7, i64 %2   
  br label %loop                                            
                                                                                                                                             
loop:                                             ; preds = %loop, %entry
  %loop_var = phi i64 [ 0, %entry ], [ %"loop_var+1", %loop ]                                                                                
  %scevgep8 = getelementptr double, double* %scevgep, i64 %loop_var
  %a = load double, double* %scevgep8, align 8      
  %4 = fcmp olt double %a, 1.000000e+03                   
  %5 = sext i1 %4 to i8                             
```

### CODE REVIEW

File: cpp/src/gandiva/arrow.h [hunk 1]
```diff
@@ -25,6 +25,7 @@
 #include <arrow/builder.h>
 #include <arrow/pretty_print.h>
 #include <arrow/record_batch.h>
+#include <arrow/result.h>
```
Comment (pitrou) [inline cpp/src/gandiva/arrow.h]: Can't we use `type_fwd.h` instead?

File: cpp/src/gandiva/engine.h [hunk 2]
```diff
@@ -23,9 +23,9 @@
 #include <string>
 #include <vector>
 
-#include "arrow/status.h"
 #include "arrow/util/macros.h"
 
+#include "gandiva/arrow.h"
```
Comment (pitrou) [inline cpp/src/gandiva/engine.h]: Hmm... can we keep the includes as minimal as possible? Arrow is already slow enough to build.

File: cpp/src/gandiva/engine_llvm_test.cc [hunk 3]
```diff
@@ -18,113 +18,111 @@
...
 
 #include <gtest/gtest.h>
+#include <functional>
 #include "gandiva/llvm_types.h"
 #include "gandiva/tests/test_util.h"
 
 namespace gandiva {
 
-typedef int64_t (*add_vector_func_t)(int64_t* elements, int nelements);
+typedef int64_t (*add_vector_func_t)(int64_t*, int);
```
Comment (pitrou) [inline cpp/src/gandiva/engine_llvm_test.cc]: The original signature makes it clearer what the arguments mean.

File: python/pyarrow/gandiva.pyx [hunk 4]
```diff
@@ -149,6 +149,10 @@ cdef class Projector:
         self.pool = pool
         return self
 
+    @property
+    def ir(self):
```
Comment (pitrou) [inline python/pyarrow/gandiva.pyx]: Call this `llvm_ir`?

File: python/pyarrow/tests/test_gandiva.py [hunk 5]
```diff
@@ -46,6 +46,8 @@ def test_tree_exp_builder():
     projector = gandiva.make_projector(
         schema, [expr], pa.default_memory_pool())
 
+    assert projector.ir.find("@expr_") != -1
```
Comment (pitrou) [inline python/pyarrow/tests/test_gandiva.py]: Please add a comment explaining why this assertion should hold.

File: python/pyarrow/tests/test_gandiva.py [hunk 6]
```diff
@@ -96,6 +98,7 @@ def test_filter():
     condition = builder.make_condition(cond)
 
     filter = gandiva.make_filter(table.schema, condition)
+    assert filter.ir.find("@expr_") != -1
```
Comment (pitrou) [inline python/pyarrow/tests/test_gandiva.py]: Same here.

**Review by pitrou — COMMENTED** [review]
LGTM in general, a few comments.

File: cpp/src/gandiva/arrow.h (hunk 1 above)
Comment (fsaintjacques) [reply cpp/src/gandiva/arrow.h]: I tried removing the others and minimizing but it unfolded in too many changes.

File: cpp/src/gandiva/engine.h (hunk 2 above)
Comment (fsaintjacques) [reply cpp/src/gandiva/engine.h]: clangd has this option to automatically add header when auto-completing. I disabled it.

**Review by projjal — COMMENTED** [review]
Looks good to me.

File: cpp/src/gandiva/llvm_generator_test.cc [hunk 7]
```diff
@@ -40,22 +41,18 @@ class TestLLVMGenerator : public ::testing::Test {
...
-  auto status = LLVMGenerator::Make(TestConfiguration(), &generator);
-  EXPECT_TRUE(status.ok()) << status.message();
+  ASSERT_OK(LLVMGenerator::Make(TestConfiguration(), &generator));
 
   llvm::Module* module = generator->module();
   for (auto& iter : registry_) {
-    llvm::Function* fn = module->getFunction(iter.pc_name());
-    EXPECT_NE(fn, nullptr) << "function " << iter.pc_name()
-                           << " missing in precompiled module\n";
+    EXPECT_NE(module->getFunction(iter.pc_name()), nullptr);
```
Comment (pravindra) [inline cpp/src/gandiva/llvm_generator_test.cc]: having the name in stderr is helpful for debugging.

**Review by pravindra — APPROVED** [review]

File: cpp/src/gandiva/llvm_generator_test.cc (hunk 7 above)
Comment (fsaintjacques) [reply cpp/src/gandiva/llvm_generator_test.cc]: I recommend to set this in your `.gdbinit`

```
set environment GTEST_BREAK_ON_FAILURE=1
```

Then you run the failing test under gdb, it'll break at the first failing test.

Comment (wesm) [reply cpp/src/gandiva/llvm_generator_test.cc]: He might mean that the problem may be more evident from looking at test log files :) I think it's OK for now, if it becomes an issue we can improve

Comment (pravindra) [reply cpp/src/gandiva/llvm_generator_test.cc]: yes, wes - I meant easier to get error from log files or travis output !. but, we'll only hit this when adding new functions - so, it's obvious anyway.

**Pull request closed without merging**